### ISSUE TITLE
JWK response without "kid" parameter is not accepted

### ISSUE THREAD

**Issue by ucyo** [boylesoftware/tomcat-oidcauth#20]
Hi there,

during the configuration using `jwks` the module expects the `kid` parameter.
Unfortunately our jwks setup does not return a `kid` parameter:

[https://login.helmholtz-data-federation.de/oauth2/jwk](https://login.helmholtz-data-federation.de/oauth2/.well-known/openid-configuration)

Is there a workaround for this?

Error message of tomcat is:
```
[..]
 Caused by: java.lang.IllegalArgumentException: JSONObject["kid"] not found.
        at org.bsworks.util.json.JSONObject.get(JSONObject.java:381)
 	at org.bsworks.util.json.JSONObject.getString(JSONObject.java:582)
 	at org.bsworks.catalina.authenticator.oidc.JWKSet.<init>(JWKSet.java:56)
 	at org.bsworks.catalina.authenticator.oidc.OPConfiguration$1.parseDocument(OPConfiguration.java:63)
 	at org.bsworks.catalina.authenticator.oidc.OPConfiguration$1.parseDocument(OPCon
[..]
```

**Comment by levahim** [comment]
How is the key referred to in the JWTs? Can you provide an example of your JWT? Normally, in the JWT header, if RSA is used for JWT signature verification, there's `kid` property used to select the key from the JWK set.

**Comment by ucyo** [comment]
Here is one example of the JWT provided by [jwiks](https://login.helmholtz-data-federation.de/oauth2/jwk):

```
{"keys":[{"kty":"RSA","e":"AQAB","use":"sig","n":"l3R3...3UB28"}]}
```
I don't know enough about the protocols used by OpenID Connect to answer your question. The [documentation](https://tools.ietf.org/html/rfc7515#section-4.1.4) says that it is being used, as you mentioned, to select the appropriate key or signal a key change. Does this maybe hint that there is a single default key which is being used?

**Comment by levahim** [comment]
I really have to see a JWT to answer that. It can also be called an ID Token. I need to see how it is signed.

**Comment by ucyo** [comment]
I contacted the administrators and will get back to you when I have more information...

**Comment by levahim** [comment]
If there are some other apps in your env that use the same OP for user login, perhaps you can see the network traffic in the browser and extract the token from there. Looks like a long base64 string consisting of three parts separated with dots.
Since you're asking the admins, would be also curious to find out what OP implementation they use (or is it custom built?).

**Comment by ucyo** [comment]
I just got feedback from the admin and he suggests using an empty string `""` for `kid`.

> I really have to see a JWT to answer that. It can also be called an ID Token. I need to see how it is signed.

I tried using the Inspector of Chrome to get the traffic and token, but didn't succeed. A webpage using the backend I want to implement is https://login.helmholtz-data-federation.de/home/. There is the option to use GitHub for authentication. Maybe you'll have more success using the Inspector for catching the token?

**Comment by levahim** [comment]
Your admin's suggestion is... useless.
GitHub uses OAuth 2.0, but it does not use OpenID Connect, nor it utilizes JWT id tokens.
I'm going to create branch and disable the requirement for the `kid`. You can build from that branch and run your app. It will probably fail, but I'll need to see the debug log--the authenticator logs the JWT. Make sure you enable FINE debug lever for `org.bsworks` (e.g. `org.bsworks.level = FINE` in your `logging.properties`, if that's what you use). Watch for "decoded ID token" in the log.

**Comment by levahim** [comment]
Find the code on "no-kid-jwks" branch. Let me know how it goes...

**Comment by levahim** [comment]
This has been resolved in version 2.3.1.
